src/native-shim.js

```diff
@@ -18,15 +18,125 @@
  * shim must only be loaded in browsers that do.
  */
 (() => {
-  let origHTMLElement = HTMLElement;
-  // TODO(justinfagnani): Tests!!
+  'use strict';
+
+  const NativeHTMLElement = HTMLElement;
+  const nativeDefine = window.customElements.define;
+  const nativeGet = window.customElements.get;
+
+  /**
+   * Map of user-provided constructors to tag names.
+   *
+   * @type {Map<Function, string>}
+   */
+  const tagnameByConstructor = new Map();
+
+  /**
+   * Map of tag anmes to user-provided constructors.
+   *
+   * @type {Map<string, Function>}
+   */
+  const constructorByTagname = new Map();
+
+
+  // /**
+  //  * Whether the constructors are being called by a browser process, ie parsing
+  //  * or createElement.
+  //  */
+  // let browserConstruction = false;
```

Comment:
Commented out so I can expose to test some pre and post conditions. I'll move back to these private variables soon.